p2panda-net/src/sync/accept.rs

```diff
@@ -0,0 +1,221 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+use std::sync::Arc;
+
+use anyhow::Result;
+use futures_util::{AsyncRead, AsyncWrite, SinkExt};
+use iroh_net::key::PublicKey;
+use p2panda_sync::{FromSync, SyncError, SyncProtocol, Topic};
+use tokio::sync::{mpsc, oneshot};
+use tokio::task::JoinHandle;
+use tokio_util::sync::PollSender;
+use tracing::{debug, error};
+
+use crate::engine::ToEngineActor;
+use crate::TopicId;
+
+/// Accept a sync protocol session over the provided bi-directional stream for the given peer and
+/// topic.
+pub async fn accept_sync<T, S, R>(
+    mut send: &mut S,
+    mut recv: &mut R,
+    peer: PublicKey,
+    sync_protocol: Arc<dyn for<'a> SyncProtocol<'a, T> + 'static>,
+    engine_actor_tx: mpsc::Sender<ToEngineActor<T>>,
+) -> Result<(), SyncError>
+where
+    T: Topic + TopicId + 'static,
+    S: AsyncWrite + Send + Unpin,
+    R: AsyncRead + Send + Unpin,
+{
+    debug!("accept sync session with peer {}", peer);
+
+    engine_actor_tx
+        .send(ToEngineActor::SyncStart { topic: None, peer })
+        .await
+        .map_err(|err| {
+            SyncError::Critical(format!("engine_actor_tx failed sending sync start: {err}"))
+        })?;
+
+    // Set up a channel for receiving messages from the sync session.
+    let (tx, mut rx) = mpsc::channel::<FromSync<T>>(128);
+    let mut sink = PollSender::new(tx).sink_map_err(|e| SyncError::Critical(e.to_string()));
+
+    // Set up a channel for sending over errors to the "glue" task which occurred during sync.
+    let (sync_error_tx, mut sync_error_rx) = oneshot::channel::<SyncError>();
+
+    // Spawn a "glue" task which represents the layer between the sync session and the engine.
+    //
+    // It picks up any messages from the sync session makes sure that the "Two-Phase Sync Flow" is
+    // followed (I. "Handshake" Phase & II. "Data Sync" Phase) and the engine accordingly informed
+    // about it.
+    //
+    // If the task detects any invalid behaviour of the sync flow, it fails critically, indicating
+    // that the sync protocol implementation does not behave correctly and is not compatible with
+    // the engine.
+    //
+    // Additionally, the task forwards any synced application data straight to the engine.
+    let glue_task_handle: JoinHandle<Result<(), SyncError>> = tokio::spawn(async move {
+        let mut topic = None;
+
+        loop {
+            tokio::select! {
+                biased;
+
+                Ok(err) = &mut sync_error_rx => {
```

Comment:
This was moved up again in the `select!` priority. We need to catch errors on the oneshot channel, otherwise we can never propagate the error further.